File: Python/day20.py
#!/usr/bin/env python3
from typing import Tuple, Dict

from util import load

Coord = Tuple[int, int]
Algorithm = Dict[int, bool]
Grid = Dict[Coord, int]

DIM = 0


def new_value(g: Grid, algo: Algorithm, c: Coord, default: int) -> int:
    x, y = c
    idx = 0
    for dy in [-1, 0, 1]:
        for dx in [-1, 0, 1]:
            nc = (x + dx, y + dy)
            idx = idx * 2 + g.get(nc, algo[default])
    return algo[idx]


def evolve_1(g: Grid, algo: Algorithm, delta, default) -> Grid:
    ng = {}
    search_delta = delta + 1
    for y in range(-search_delta, DIM + search_delta):
        for x in range(-search_delta, DIM + search_delta):
            lit = new_value(g, algo, (x, y), default)
            ng[(x, y)] = lit
    return ng


def evolve(g: Grid, algo: Algorithm, steps: int) -> Grid:
    delta = 0
    default = 0 if algo[0] == 0 else 511
    for i in range(steps):
        print(f'Lit pixels after {i} enhancements = {lit_pixels(g)}, '
              f'with delta {delta} and default {default}')
        # display(g, delta)
        g = evolve_1(g, algo, delta, default)
        delta += 1
        default = 511 if algo[default] else 0
    return g


def lit_pixels(g: Grid) -> int:
    return sum(filter(lambda p: p == 1, g.values()))


def display(g: Grid, delta: int):
    ls = [
        [('#' if g.get((x, y), 0) else '.')
         for x in range(-delta, DIM + delta)]
        for y in range(-delta, DIM + delta)]
    print(f'Delta = {delta}')
    for l in ls:
        print(''.join(l))


if __name__ == "__main__":
    lines = load('day20.txt')
    algorithm: Algorithm = {
        i: (1 if c == '#' else 0) for i, c in enumerate(lines[0])}
    grid: Grid = {
        (x, y): (1 if c == '#' else 0)
        for y, line in enumerate(lines[2:])
        for x, c in enumerate(line)
    }
    DIM = len(lines[2])

    result = evolve(grid, algorithm, 2)
    num_lit = lit_pixels(result)
    print(f'Lit pixels after 2 enhancements = {num_lit}')
    assert num_lit == 5464

    result = evolve(grid, algorithm, 50)
    num_lit = lit_pixels(result)
    print(f'Lit pixels after 50 enhancements = {num_lit}')
    assert num_lit == 19228

File: Elixir/lib/day20.ex
defmodule Day20 do
  @moduledoc """
  Documentation for `Day20`.
  """

  @dim 100

  def lit(c), do: (if c == "#", do: 1, else: 0)

  def input() do
    [rules | lines] = File.read!("data/day20.txt")
                      |> String.split("\n")
    {
      rules
      |> String.split("", trim: true)
      |> Stream.map(&lit/1)
      |> Stream.with_index()
      |> Enum.reduce(
           %{},
           fn {l, i}, acc ->
             Map.put(acc, i, l)
           end
         ),
      lines
      |> Stream.filter(&(String.length(&1) > 0))
      |> Stream.with_index()
      |> Enum.reduce(
           %{},
           fn {line, y}, acc ->
             String.split(line, "", trim: true)
             |> Stream.with_index()
             |> Enum.reduce(
                  acc,
                  fn {c, x}, acc ->
                    Map.put(acc, {x, y}, lit(c))
                  end
                )
           end
         )
    }
  end

  def new_value(image, algo, {x, y}, default) do
    algo[
        -1..1
        |> Enum.reduce(
             0,
             fn dy, idx ->
               -1..1
               |> Enum.reduce(
                    idx,
                    fn dx, idx ->
                      idx * 2 + Map.get(image, {x + dx, y + dy}, algo[default])
                    end
                  )
             end
           )
        ]
  end

  def evolve(image, algo, delta, default) do
    d = delta + 1
    -d..(@dim + d - 1)
    |> Enum.reduce(
         %{},
         fn y, acc ->
           -d..(@dim + d - 1)
           |> Enum.reduce(
                acc,
                fn x, acc ->
                  Map.put(acc, {x, y}, new_value(image, algo, {x, y}, default))
                end
              )
         end
       )
  end

  def evolve(image, _algo, _delta, _default, 0), do: image
  def evolve(image, algo, delta, default, generation) do
    image = evolve(image, algo, delta, default)
    evolve(image, algo, delta + 1, (if algo[default] == 1, do: 511, else: 0), generation - 1)
  end

  def evolve(generations) do
    {algo, image} = input()
    delta = 0
    {default, _} = algo
                   |> Map.to_list()
                   |> Stream.filter(fn {_idx, lit} -> lit == 0 end)
                   |> Enum.at(0)
    evolve(image, algo, delta, default, generations)
  end

  def lit_pixels(image) do
    image
    |> Map.values()
    |> Enum.filter(fn v -> v == 1 end)
    |> Enum.count()
  end
end
